src/arch/riscv/plic.c

```diff
@@ -1,90 +1,123 @@
 #include <arch/riscv/plic.h>
 #include <arch/riscv/sbi.h>
 #include <nautilus/cpu.h>
-#include <nautilus/devicetree.h>
 #include <nautilus/naut_types.h>
 #include <nautilus/percpu.h>
+#include <nautilus/fdt.h>
+
+typedef struct plic_context {
+  off_t enable_offset;
+  off_t context_offset;
+} plic_context_t;
 
 static addr_t plic_addr = 0;
 
-#define MREG(x) *((uint32_t *)x)
+static plic_context_t *contexts = NULL;
+// static plic_context_t contexts[100];
 
 #define PLIC plic_addr
-#define PLIC_PRIORITY MREG(PLIC + 0x0)
-#define PLIC_PENDING MREG(PLIC + 0x1000)
-#define PLIC_MENABLE(hart) MREG(PLIC + 0x2000 + (hart)*0x100)
-#define PLIC_SENABLE(hart) MREG(PLIC + 0x2080 + (hart)*0x100)
-#define PLIC_MPRIORITY(hart) MREG(PLIC + 0x201000 + (hart)*0x2000)
-#define PLIC_SPRIORITY(hart) MREG(PLIC + 0x200000 + (hart)*0x2000)
-#define PLIC_MCLAIM(hart) MREG(PLI + 0x201004 + (hart)*0x2000)
-#define PLIC_SCLAIM(hart) MREG(PLIC + 0x200004 + (hart)*0x2000)
-
-#define ENABLE_BASE 0x2000
-#define ENABLE_PER_HART 0x100
-
-bool_t dtb_node_plic_compatible(struct dtb_node *n) {
-    for (off_t i = 0; i < n->ncompat; i++) {
-        if (strstr(n->compatible[i], "plic")) {
-            return true;
+#define MREG(x) *((uint32_t *)(PLIC + (x)))
+#define READ_REG(x) *((uint32_t *)((x)))
+
+#define PLIC_PRIORITY_BASE 0x000000U
+
+#define PLIC_PENDING_BASE 0x1000U
+
+#define PLIC_ENABLE_BASE 0x002000U
+#define PLIC_ENABLE_STRIDE 0x80U
+#define IRQ_ENABLE 1
+#define IRQ_DISABLE 0
+
+#define PLIC_CONTEXT_BASE 0x200000U
+#define PLIC_CONTEXT_STRIDE 0x1000U
+#define PLIC_CONTEXT_THRESHOLD 0x0U
+#define PLIC_CONTEXT_CLAIM 0x4U
+
+#define PLIC_PRIORITY(n) (PLIC_PRIORITY_BASE + (n) * sizeof(uint32_t))
+#define PLIC_ENABLE(n, h) (contexts[h].enable_offset + ((n) / 32) * sizeof(uint32_t))
+#define PLIC_THRESHOLD(h) (contexts[h].context_offset + PLIC_CONTEXT_THRESHOLD)
+#define PLIC_CLAIM(h) (contexts[h].context_offset + PLIC_CONTEXT_CLAIM)
+
+void plic_init(unsigned long fdt) {
+    int offset = fdt_node_offset_by_compatible(fdt, -1, "sifive,plic-1.0.0");
+    if (offset < 0) {
+        // something is bad, no plic found
+        return;
+    }
+    off_t addr = fdt_getreg_address(fdt, offset);
+    PLIC = addr;
+
+    int lenp = 0;
+    void *ints_extended_prop = fdt_getprop(fdt, offset, "interrupts-extended", &lenp);
+    if (ints_extended_prop != NULL) {
+        uint32_t *vals = (uint32_t *)ints_extended_prop;
+        int context_count = lenp / 8;
+
+        contexts = kmem_malloc(context_count * sizeof(plic_context_t));
```

Comment:
This is probably not ideal since it requires malloc working before the plic init, is there a better way?

Reply:
I think it's fine to allocate before interrupts. The allocator is guarded by a spinlock instead of something fancy, so I don't see why it would be a problem.